relentless/potential/pair.py

```diff
@@ -156,6 +155,298 @@ def _derivative(self, param, r, epsilon, sigma, **params):
+
+        Parameters
+        ----------
+        pair : array_like
+            The pair for which to calculate the derivative.
+        param : str
+            The parameter with respect to which to take the derivative.
+        r : array_like
+            The location(s) at which to calculate the derivative.
+
+        Returns
+        -------
+        scalar or array_like
+            The derivative at the specified location(s).
+            The returned quantity will be a scalar if `r` is scalar
+            or a numpy array if `r` is array_like.
+
+        Raises
+        ------
+        TypeError
+            If an r value is not of type :py:class:`Variable`
+        TypeError
+            If a knot value is not of type :py:class:`Variable`
+
+        """
+        for i in range(self.num_knots):
+            ri,ki = self._knot_params(i)
+            if not isinstance(self.coeff[pair][ri], core.Variable):
+                raise TypeError('All r values must be Variables')
+            if not isinstance(self.coeff[pair][ki], core.Variable):
+                raise TypeError('All knot values must be Variables')
+        return super().derivative(pair, param, r)
+
+    def _derivative(self, param, r, params, **pars):
+        """Evaluates the spline parameter derivative using finite differencing.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        r : array_like
+            The value or values of r at which to evaluate the energy.
+        params : dict
+            The parameters (r and knot values) for the potential function.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the derivative evaluated at r.
+
+        Raises
+        ------
+        ValueError
+            If the parameter with respect to which to take the derivative is not a knot value.
+
+        """
+        if 'knot' not in param:
+            raise ValueError('Parameter derivative can only be taken for knot values')
+
+        r,d,s = self._zeros(r)
+        h = 0.001
+
+        #perturb knot param value
+        knot_p = params[param].value
+        params[param].value = knot_p + h
+        f_high = self._interpolate(params)(r)
+        params[param].value = knot_p - h
+        f_low = self._interpolate(params)(r)
+
+        params[param].value = knot_p
+        d = (f_high - f_low)/(2*h)
+        if s:
+            d = d.item()
+        return d
+
+    def _interpolate(self, params):
+        """Interpolates the knot points into a continuous spline potential.
+
+        Parameters
+        ----------
+        params : array_like
+            The array of knot values.
+
+        Returns
+        -------
+        :py:class:`Interpolator`
+            The interpolated spline potential.
+
+        """
+        r = np.empty(self.num_knots)
+        u = np.empty(self.num_knots)
```

Comment:
I would recommend using `zeros`, just to be safe. `empty` gives you whatever bits are in the memory when it's allocated, which are undefined values.